Tuning documentation for foreman installer (#1504)
theforeman/theforeman.org: plugins/katello/3.14/installation/index.md

        reverse: true
        contact: hostmaster.example.com.

## Tuning options
The Foreman installer supports automatic tuning of your environment using predefined tuning profiles.  These tuning profiles are the result of a culmination of extensive learning from Foreman environments deployed at scale in large user environments.

> When the foreman-installer is run, it is deployed with a predefined tuning profile named `default`. The definition of the `default` profile can be found in `/usr/share/foreman-installer/config/foreman.hiera/tuning/common.yaml`.

Reviewer: This may be a bit misleading. As it's written now it implies when `medium` is selected, `common` is no longer applied which isn't true. Common is used for all tuning. In a mathematical way: `tuning = common + profile`. Technically Hiera looks at `foreman.hiera/tuning/sizes/default.yaml` and then `foreman.hiera/tuning/common.yaml`. When a file doesn't exist, Hiera just continues.

Reviewer 2: This is true for now unless we make an empty default yaml file. So for me it reads OK.

Reviewer: I disagree. `common` is also used with `medium`. The current text implies you **only** need to look at it when you use default. Given later on we recommend users to look at their tuning, they may keep redundant tuning in `custom-hiera.yaml`.

I'd split it up and leave out how to find the exact tuning till later.

```suggestion
When the foreman-installer is run, it is deployed with a predefined tuning profile named `default`. Other than the default tuned profile, foreman-installer supports 4 different tuning profiles:
```

Reviewer 2: I kinda disagree with that idea as this goes back to my whole problem with the naming in the first place. The default tuning profile is common. Common is the default. Common is *always* applied so it is the defaults. When looking at say medium, I should be pointed at the medium profile as thats what makes medium _different_ than the default.

Author: I have a design question on the real necessity of `foreman.hiera/tuning/common.yaml` file.
I think we should do one of the two options:
Option 1
**********
1. Strip out the changing parameters of  `foreman.hiera/tuning/common.yaml` and move them to a new `foreman.hiera/tuning/sizes/default.yaml` file.
2. And mention in docs that tuning will be always applied in this order:
   - common.yaml
   - `selected tuning profile`. For e.g., medium.yaml  (Note any configuration here overrides default .yaml)
   - custom-hiera.yaml always precedes all other tuning profiles above.

Option 2
**********
- Rename `foreman.hiera/tuning/common.yaml` to `default.yaml` and move it to `foreman.hiera/tuning/sizes`. Note: No common.yaml anymore.
- And mention in docs that tuning will be always applied in this order:
   - default.yaml
   - `selected tuning profile`. For e.g., medium.yaml  (Note any configuration here overrides default .yaml)
   - custom-hiera.yaml always precedes all other tuning profiles above.

Other than the default tuned profile, foreman-installer supports 4 different tuning profiles:

  * medium
  * large
  * extra-large
  * extra-extra-large

> Definitions of various tuning profiles can be found in this directory `/usr/share/foreman-installer/config/foreman.hiera/tuning/sizes/`.  Based on your environment needs, use one of the tuning profiles (`medium`, `large`, `extra-large`, `extra-extra-large`) in the installer.  For example, `medium` profile can be applied like:

Reviewer: ```suggestion
Based on your environment needs, use one of the tuning profiles (`medium`, `large`, `extra-large`, `extra-extra-large`) in the installer.  For example, `medium` profile can be applied like:
```

```bash
foreman-installer --tuning medium
```

To reset to the default profile:

```bash
foreman-installer --tuning default
```

> **Note**

Reviewer: I'd move all these notes to the bottom of the section, below the table. My reasoning is that docs should go from most relevant to least relevant. Using `custom-hiera.yaml` is unsupported and you are expected to be an advanced user already. That means it should be at the bottom.

Author: Good point about not emphasizing too much on `custom-hiera.yaml` for entry-level. Now I created a new sub-section.  Moved the note on custom-hiera.yaml to the end of the first section (not to the very end as per your comment) because the second sub-section is mostly for choosing the right tuned profile.

- Using the `--tuning` option does not update `/etc/foreman-installer/custom-hiera.yml`, instead it directly updates the required configuration as specified in the corresponding tuning profile. You can still use `custom-hiera.yml` to override any configuration if really needed.

> - If you had already used `custom-hiera.yml` and starting to use the tuned profiles, you may want to review the definition of tuned profiles (`/usr/share/foreman-installer/config/foreman.hiera/tuning/sizes/`) and remove the duplicated configuration entries from your `custom-hiera.yml`.

Reviewer: ```suggestion
- If you had already used `custom-hiera.yml` and starting to use the tuned profiles, you may want to review the definition of tuned profiles (`/usr/share/foreman-installer/config/foreman.hiera/tuning/common.yaml` and `/usr/share/foreman-installer/config/foreman.hiera/tuning/sizes/`) and remove the duplicated configuration entries from your `custom-hiera.yml`.
```

```

Sample output for `default` tuning:

```bash
foreman-installer --help | grep tuning
    --tuning INSTALLATION_SIZE    Tune for an installation size. Choices: default, medium, large, extra-large, extra-extra-large (default: "default")
```

**Note**

- Definitions of various tuning profiles can be found in this directory `/usr/share/foreman-installer/config/foreman.hiera/tuning/sizes/`.  Note that `common.yaml` is always applied and the selected tuning profile (e.g., `medium`) is applied on top and takes precedence.
- Using the `--tuning` option does not update `/etc/foreman-installer/custom-hiera.yml`, instead it directly updates the required configuration as specified in the corresponding tuning profile. You can still use `custom-hiera.yml` to override any configuration if really needed.
- If you had already used `custom-hiera.yml` and starting to use the tuned profiles, you may want to review the definition of tuned profiles (`/usr/share/foreman-installer/config/foreman.hiera/tuning/common.yaml` and `/usr/share/foreman-installer/config/foreman.hiera/tuning/sizes/`) and remove the duplicated configuration entries from your `custom-hiera.yml`.

> - You can also optionally use `foreman-installer --noop` to run the installer in a test mode and identify what configurations will be changed before actually running the installer.

Reviewer: To me it's clear that this can also be combined with `--tuning <profile>` and I also know that `--noop` implies not saving the answers (including tuning). Is that also clear for the average user?

Author: ah good point, I will update this.

Author: done

> ### Which tuning profile should you choose?

Reviewer: :+1: I like an explicit section here

> It is difficult to find the exact tuning profile for a specific environment in the first attempt because it depends on various factors like the number of managed hosts, the features used in scale (E.g., Remote Execution), the bulk action on hosts, the total amount of content, amount of host traffic to foreman, etc.  Our recommendation is that you start with the tuning profile guidance as shown in the below table based on the number of managed hosts and scale up your environment as needed.

Reviewer: `the bulk action` looks odd. Did you mean `bulk actions`?
```suggestion
It is difficult to find the exact tuning profile for a specific environment in the first attempt because it depends on various factors like the number of managed hosts, the features used in scale (E.g., Remote Execution), bulk actions on hosts, the total amount of content, amount of host traffic to foreman, etc.  Our recommendation is that you start with the tuning profile guidance as shown in the below table based on the number of managed hosts and scale up your environment as needed.
```

Author: done

        reverse: true
        contact: hostmaster.example.com.

## Tuning options
The Foreman installer supports automatic tuning of your environment using predefined tuning profiles.  These tuning profiles are the result of a culmination of extensive learning from Foreman environments deployed at scale in large user environments.

When the foreman-installer is run, it is deployed with a predefined tuning profile named `default`. The definition of the `default` profile can be found in `/usr/share/foreman-installer/config/foreman.hiera/tuning/common.yaml`.

Other than the default tuned profile, foreman-installer supports 4 different tuning profiles:

  * medium
  * large
  * extra-large
  * extra-extra-large

Definitions of various tuning profiles can be found in this directory `/usr/share/foreman-installer/config/foreman.hiera/tuning/sizes/`.  Based on your environment needs, use one of the tuning profiles in the installer as shown below.

```bash
foreman-installer --tuning medium
```

> ```bash

Reviewer: I'm wondering if it makes sense to specify all options since it's essentially the same.

Author: I wanted to make it super clear, but it feels redundant - I can combine them - providing just one example.

foreman-installer --tuning large
```

```bash
foreman-installer --tuning extra-large
```

```bash
foreman-installer --tuning extra-extra-large
```

To reset to the default profile:

```bash
foreman-installer --tuning default
```

**Note**

- Using the `--tuning` option does not update `/etc/foreman-installer/custom-hiera.yml`, instead it directly updates the required configuration as specified in the corresponding tuning profile. You can still use `custom-hiera.yml` to override any configuration if really needed.
- If you had already used `custom-hiera.yml` and starting to use the tuned profiles, you may want to review the definition of tuned profiles (`/usr/share/foreman-installer/config/foreman.hiera/tuning/sizes/`) and remove the duplicated configuration entries from your `custom-hiera.yml`.
- Use `foreman-installer --help | grep tuning` to identify the current tuning level.

Sample output for `default` tuning:

> ```bash

Reviewer: On Github you can use `console` as well and use `#` (or `$`, but this should run as root) to separate the command from output:
```console
# foreman-installer --help | grep tuning
     --tuning INSTALLATION_SIZE    Tune for an installation size. Choices: default, medium, large, extra-large, extra-extra-large (default: "default")
```

Don't know if our website renders it in the same way.

Author: yes I have seen it in other projects.  But I don't see any reference of it being used in this project. So I used the `bash` format.

foreman-installer --help | grep tuning
    --tuning INSTALLATION_SIZE    Tune for an installation size. Choices: default, medium, large, extra-large, extra-extra-large (default: "default")
```

Sample output for `medium` tuning:

```bash
foreman-installer --help | grep tuning
    --tuning INSTALLATION_SIZE    Tune for an installation size. Choices: default, medium, large, extra-large, extra-extra-large (default: "medium")
```

It is difficult to find the exact tuning profile for a specific environment in the first attempt because it depends on various factors like the number of managed hosts, the features used in scale (E.g., Remote Execution), the bulk action on hosts, the total amount of content, amount of host traffic to foreman, etc.  Our recommendation is that you start with the tuning profile guidance as shown in the below table based on the number of managed hosts and scale up your environment as needed.

> **Note** The information in the table below is just a guidance.  It is strongly recommended that you monitor the foreman environment regularly and tune up as required.

Reviewer: https://github.com/theforeman/foreman-installer/commit/8f8f2beaf46ed1658a28b3877a571054a5c5cec6 makes these a hard lower limits and need `--disable-system-checks` to skip. Maybe worth mentioning.

Author: good point, will add
